### PR TITLE
Buffer rendered CacheTagHelper content strings to a custom TextWriter

### LINKED ISSUE #4893: CacheTagHelper+StringBuilderContent is too inefficient
Here's some sampling data from a customer-provided app:

image

Writing out the output of the cache tags is 20% of the overall time in the application, which seems like it can be improved based on the fact that we've **already** rendered this content to text.

They are making heavy use of the `<asp-cache>` tag to avoid re-rendering some blocks of markup in their views. While we're pretty good with memory in the cache tag, we end up writing out the result character-by-character, which is inefficient.

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs#L197

We should make use of `StringBuilder.CopyTo` (with a `stackalloc char[]`) or even better replace the `StringBuilder` with something that allows us access to the underlying arrays. We don't want to use pooled memory here because we don't know how long the cached item will live.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #4893

### CODE REVIEW

**Comment by pranavkm** [comment]
I must've hit a degenerate case of `StringBuilder[this]` when I tried this with https://github.com/aspnet/Performance/tree/dev/testapp/LargePageMvc. 
For 2k requests:

### Before
image

### After
image

Revived`StringCollectionTextWriter` for this

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedCharBuffer.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        public ICharBufferSource BufferSource { get; }
+
+        public IList<char[]> Pages { get; } = new List<char[]>();
+
+        public char[] CurrentPage { get; private set; }
+
+        // The next 'free' character
+        public int CharIndex { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedCharBuffer.cs]: Consider making this 'length' and just doing mod in some places.

**Comment by pranavkm** [comment]
### Before
image

### After
image

**Comment by pranavkm** [comment]
:up: :date:

**Pull request closed without merging**